File: scripts/issueTracker.py
# This is python 3 script.
# pip install requests pyyaml termcolor
# Checks files for existing issues, checks github for latest comment
# Make sure to run `export GITHUB_TOKEN=your_token_here` prior to running the script

import requests
import yaml
import os
import time
from datetime import datetime, timedelta
from termcolor import colored

# Constants
TIME_THRESHOLD_DAYS = 180
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
DIRECTORIES = ["_android", "_iphone", "_hardware", "_bearer", "_desktop"]
RETRY_COUNT = 3
RATE_LIMIT_SLEEP = 60  # Wait for 60 seconds on rate limit

# Get GitHub token from environment variable
GITHUB_TOKEN = os.getenv('GITHUB_TOKEN')

if not GITHUB_TOKEN:
    raise ValueError("GitHub token not found. Please set the GITHUB_TOKEN environment variable.")

def get_headers():
    return {
        'Authorization': f'token {GITHUB_TOKEN}',
        'Accept': 'application/vnd.github.v3+json'
    }

def get_issues_from_markdown():
    issues = []

    for directory in DIRECTORIES:
        dir_path = os.path.join(BASE_DIR, directory)
        if os.path.exists(dir_path):
            for filename in os.listdir(dir_path):
                if filename.endswith(".md"):
                    filepath = os.path.join(dir_path, filename)
                    with open(filepath, 'r') as file:
                        content = file.read()
                        front_matter = yaml.safe_load(content.split('---')[1])
                        if 'issue' in front_matter and front_matter['issue']:
                            issues.append({
                                "app_id": filename,
                                "issue_url": front_matter['issue']
                            })

    return issues

def fetch_url(url, retry_count=RETRY_COUNT):
    for attempt in range(retry_count):
        try:
            response = requests.get(url, headers=get_headers())
            if response.status_code == 200:
                return response.json()
            elif response.status_code == 429:  # Rate limit error
                print(colored(f"Rate limit exceeded. Waiting for {RATE_LIMIT_SLEEP} seconds.", "yellow"))
                time.sleep(RATE_LIMIT_SLEEP)
            else:
                print(colored(f"Failed to fetch URL: {url}. Status code: {response.status_code}. Response: {response.text}", "red"))
        except requests.RequestException as e:
            print(colored(f"Request exception for URL: {url}. Error: {e}", "red"))
    return None

def fetch_issue_details(issue_url):
    if not issue_url.startswith("https://github.com/") and not issue_url.startswith("https://gitlab.com/"):
        print(colored(f"Skipping invalid URL: {issue_url}", "red"))
        return None

    try:
        parts = issue_url.split('/')
        if "github.com" in issue_url:
            owner = parts[3]
            repo = parts[4]
            issue_number = parts[6]
            issue_api_url = f"https://api.github.com/repos/{owner}/{repo}/issues/{issue_number}"
            comments_api_url = f"https://api.github.com/repos/{owner}/{repo}/issues/{issue_number}/comments"
        elif "gitlab.com" in issue_url:
            owner = parts[3]
            repo = parts[4]
            issue_number = parts[6]
            issue_api_url = f"https://gitlab.com/api/v4/projects/{owner}%2F{repo}/issues/{issue_number}"
            comments_api_url = f"https://gitlab.com/api/v4/projects/{owner}%2F{repo}/issues/{issue_number}/notes"
        else:
            print(colored(f"Skipping unsupported provider URL: {issue_url}", "red"))
            return None

        issue_details = fetch_url(issue_api_url)
        if not issue_details:
            return None

        comments = fetch_url(comments_api_url)
        if comments is None:
            return None

        last_comment_date = comments[-1]['updated_at'] if comments else issue_details['created_at']

        return {
            "app_id": issue_details.get('title', 'N/A'),
            "issue_url": issue_url,
            "date_issue_created": issue_details['created_at'],
            "date_last_comment": last_comment_date
        }
    except IndexError:
        print(colored(f"Error parsing URL: {issue_url}. The URL format seems incorrect.", "red"))
    except Exception as e:
        print(colored(f"Unexpected error for URL: {issue_url}. Error: {e}", "red"))
    return None

def filter_and_sort_issues(issues):
    filtered_issues = []
    threshold_date = datetime.now() - timedelta(days=TIME_THRESHOLD_DAYS)

    for issue in issues:
        issue_details = fetch_issue_details(issue['issue_url'])
        if issue_details:
            last_comment_date = datetime.strptime(issue_details['date_last_comment'], "%Y-%m-%dT%H:%M:%SZ")

            if last_comment_date < threshold_date:
                filtered_issues.append(issue_details)

    sorted_issues = sorted(filtered_issues, key=lambda x: x['date_last_comment'])
    return sorted_issues

def output_results(issues):
    print(f"{'App ID':<20} | {'Issue URL':<50} | {'Date Issue Created':<25} | {'Date Last Comment':<25}")
    print("-" * 125)
    for issue in issues:
        print(f"{issue['app_id']:<20} | {colored(issue['issue_url'], 'cyan'):<50} | {issue['date_issue_created']:<25} | {issue['date_last_comment']:<25}")

if __name__ == "__main__":
    issues = get_issues_from_markdown()
    sorted_issues = filter_and_sort_issues(issues)
    output_results(sorted_issues)

File: scripts/issueTracker.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const yaml = require('js-yaml');
const chalk = require('chalk');

const TIME_THRESHOLD_DAYS = 180;
const BASE_DIR = path.resolve(__dirname, '..');
const DIRECTORIES = ["_android", "_iphone", "_hardware", "_bearer", "_desktop"];
const RETRY_COUNT = 3;
const RATE_LIMIT_SLEEP = 60000; 

const GITHUB_TOKEN = process.env.GITHUB_TOKEN;

if (!GITHUB_TOKEN) {
    throw new Error("GitHub token not found. Please set the GITHUB_TOKEN environment variable.");
}

function getHeaders() {
    return {
        'Authorization': `token ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json'
    };
}

async function fetchUrl(url, headers, retryCount = RETRY_COUNT) {
    for (let attempt = 0; attempt < retryCount; attempt++) {
        try {
            const response = await axios.get(url, { headers });
            return response.data;
        } catch (error) {
            if (error.response && error.response.status === 429) {
                console.log(chalk.yellow(`Rate limit exceeded. Waiting for ${RATE_LIMIT_SLEEP / 1000} seconds.`));
                await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_SLEEP));
            } else {
                console.log(chalk.red(`Failed to fetch URL: ${url}. Status code: ${error.response ? error.response.status : 'unknown'}. Response: ${error.response ? JSON.stringify(error.response.data) : 'unknown'}`));
                return null;
            }
        }
    }
    return null;
}

async function fetchIssueDetails(issueUrl, appId) {
    console.log(`Evaluating App ID: ${appId}`);

    if (!issueUrl.startsWith("https://github.com/") && !issueUrl.startsWith("https://gitlab.com/")) {
        console.log(chalk.red(`Skipping invalid URL: ${issueUrl}`));
        return null;
    }

    try {
        const parts = issueUrl.split('/');
        let issueApiUrl, commentsApiUrl, headers = {};

        if (issueUrl.includes("github.com")) {
            const [owner, repo, issueNumber] = [parts[3], parts[4], parts[6]];
            issueApiUrl = `https://api.github.com/repos/${owner}/${repo}/issues/${issueNumber}`;
            commentsApiUrl = `https://api.github.com/repos/${owner}/${repo}/issues/${issueNumber}/comments`;
            headers = getHeaders();
        } else if (issueUrl.includes("gitlab.com")) {
            const [owner, repo, issueNumber] = [parts[3], parts[4], parts[parts.length - 1]];
            const projectId = encodeURIComponent(`${owner}/${repo}`);
            issueApiUrl = `https://gitlab.com/api/v4/projects/${projectId}/issues/${issueNumber}`;
            commentsApiUrl = `https://gitlab.com/api/v4/projects/${projectId}/issues/${issueNumber}/notes`;
        } else {
            console.log(chalk.red(`Skipping unsupported provider URL: ${issueUrl}`));
            return null;
        }

        const issueDetails = await fetchUrl(issueApiUrl, headers);
        if (!issueDetails) return null;

        const comments = await fetchUrl(commentsApiUrl, headers);
        if (!comments) return null;

        const lastCommentDate = comments.length ? comments[comments.length - 1].updated_at : issueDetails.created_at;

        return {
            app_id: issueDetails.title || 'N/A',
            issue_url: issueUrl,
            date_issue_created: issueDetails.created_at,
            date_last_comment: lastCommentDate
        };
    } catch (error) {
        console.log(chalk.red(`Unexpected error for URL: ${issueUrl} for App ID: ${appId}. Error: ${error.message}`));
        return null;
    }
}

function getIssuesFromMarkdown() {
    const issues = [];
    const seenUrls = new Set();

    for (const directory of DIRECTORIES) {
        const dirPath = path.join(BASE_DIR, directory);
        for (const filename of fs.readdirSync(dirPath)) {
            const filepath = path.join(dirPath, filename);
            const content = fs.readFileSync(filepath, 'utf-8');
            const frontMatter = yaml.load(content.split('---')[1]);
            if (frontMatter.issue) {
                const issueUrl = frontMatter.issue;
                const appId = frontMatter.appId || filename;
                if (!seenUrls.has(issueUrl)) {
                    issues.push({ app_id: appId, issue_url: issueUrl });
                    seenUrls.add(issueUrl);
                }
            }
        }
    }
    return issues;
}

async function filterAndSortIssues(issues) {
    const filteredIssues = [];
    const thresholdDate = new Date(Date.now() - TIME_THRESHOLD_DAYS * 24 * 60 * 60 * 1000);

    for (const issue of issues) {
        const issueDetails = await fetchIssueDetails(issue.issue_url, issue.app_id);
        if (issueDetails) {
            const lastCommentDate = new Date(issueDetails.date_last_comment);
            if (lastCommentDate < thresholdDate) {
                filteredIssues.push(issueDetails);
            }
        }
    }

    filteredIssues.sort((a, b) => new Date(a.date_last_comment) - new Date(b.date_last_comment));
    return filteredIssues;
}

function outputResults(issues) {
    console.log(`${'App ID'.padEnd(20)} | ${'Issue URL'.padEnd(50)} | ${'Date Issue Created'.padEnd(25)} | ${'Date Last Comment'.padEnd(25)}`);
    console.log('-'.repeat(125));
    for (const issue of issues) {
        console.log(`${issue.app_id.padEnd(20)} | ${chalk.cyan(issue.issue_url.padEnd(50))} | ${issue.date_issue_created.padEnd(25)} | ${issue.date_last_comment.padEnd(25)}`);
    }
}

(async function() {
    const issues = getIssuesFromMarkdown();
    const sortedIssues = await filterAndSortIssues(issues);
    outputResults(sortedIssues);
})();
